### PR TITLE
Add IIDX AC support using textage.cc as source

### PR DESCRIPTION
**Pull request by telperion**
Pretty much what it says on the label!
image
I'm gonna use it for [a tourney I'm running in February](https://www.start.gg/tournament/ruler-or-loser/details).

## Highlights
- By importing textage.cc's JavaScript files (which contain huge tables of data about chart levels and metadata) and creating a local JSDOM, we can comfortably run queries without phoning home for every request.
- IIDX tends not to give songs unique art, so I've included a system to generate SVG files that serve as "jackets", tied to each song's version of origin by accent colors in that version's theme. (I could probably place the mix name somewhere in there, too? Maybe later)
- Songs locked behind events such as WORLD TOURISM and X-record are also labeled with flags.
image

## A minor app-wide tweak
- **Support for per-chart BPM:** A handful of IIDX songs have significantly different BPM maps depending on which chart is being played. For example, crew (whose SPH, SPA, DPH, and DPA charts play various arrangements of the base song found in the SPN/DPN slots) might need to be listed as 150, 152, 155, 170, or 176 BPM!

## Outstanding (lower-priority) work
- Title/artist transliterations. There are official ones (i.e. ticker text) and remywiki generally notes them, but where exactly on the page they're listed varies a lot. Also, players generally can identify songs in Japanese characters by version, difficulty, and BPM. (Cue the classic joke of posting a hard clear on 罠 [a] and having everyone yell "yoooooo nice Mei HC")
- Theoretically we could add tiering based on community lists, but I don't wanna clutter the app in the first PR for this haha
- Better scraping of time-based, arena, and shop unlocks. textage doesn't indicate these, so I'd have to co-ingest something like bemaniwiki to get that info.

### CODE REVIEW

**Review by noahm — COMMENTED** [review]
Truly excellent work! Nothing really blocking here, so I'll leave it to you whether you want to make more changes before merging.

On the subject of transliterations/translations, you could maybe cheat a bit and copy over any exact matches for song titles in the existing DDR or SDVX files? That would at least cover your classic case of A vs Mei 😅

File: scripts/scraping/textage.mjs
Comment (telperion) [reply scripts/scraping/textage.mjs:104]: I went down that route partially because I tried and failed to just include them directly in the import script 😤 after noticing that the DDR ingest uses JSDOM, I poked around to see what else could be done with that, and found an extremely relevant stackoverflow post hah

File: scripts/import-iidx.mjs
Comment (telperion) [reply scripts/import-iidx.mjs:334]: it actually didn't even cross my mind to use the key art directly! I might leave it up to community feedback and change it to that later. in any case, having svg fallbacks is nice because they're pretty light

**Comment by telperion** [comment]
I forgot about some time-locked charts and songs, so I added a few things for that just now - but otherwise I think we should be set!

I'll ponder translations/transliterations and more automatic unlock scraping for another PR haha

**Review by noahm — APPROVED** [review]

**Pull request merged**
